lib/matplotlib/cbook/__init__.py

```diff
@@ -2191,3 +2191,68 @@ def _unikey_or_keysym_to_mplkey(unikey, keysym):
         "next": "pagedown",  # Used by tk.
     }.get(key, key)
     return key
+
+
+@functools.lru_cache(None)
+def _make_class_factory(
+        mixin_class, fmt, attr_name=None, *, _default_base_class=None):
+    """
+    Return a function that creates picklable classes inheriting from a mixin.
+
+    After ::
+
+        factory = _make_class_factory(FooMixin, fmt, attr_name)
+        FooAxes = factory(Axes)
+
+    ``Foo`` is a class that inherits from ``FooMixin`` and ``Axes`` and **is
+    picklable** (picklability is what differentiates this from a plain call to
+    `type`).  Its ``__name__`` is set to ``fmt.format(Axes.__name__)`` and the
+    base class is stored in the ``attr_name`` attribute, if not None.
+
+    Moreover, the return value of ``factory`` is memoized: calls with the same
+    ``Axes`` class always return the same subclass.
+    """
+
+    @functools.lru_cache(None)
+    def class_factory(axes_class):
+        # The parameter is named "axes_class" for backcompat but is really just
+        # a base class; no axes semantics are used.
+        base_class = axes_class
+        # _default_base_class should go away once the deprecation elapses.
+        if base_class is None and _default_base_class is not None:
+            warn_deprecated(
+                "3.5", message="Support for passing None to class factories "
+                "is deprecated since %(since)s and will be removed "
+                "%(removal)s; explicitly pass the default {} class instead."
+                .format(_default_base_class.__name__))
+            return class_factory(_default_base_class)
+
+        class subcls(mixin_class, base_class):
+            # Better approximation than __module__ = "matplotlib.cbook".
+            __module__ = mixin_class.__module__
+
+            def __reduce__(self):
+                return (_picklable_class_constructor,
+                        (mixin_class, fmt, attr_name,
+                         _default_base_class, base_class),
+                        self.__getstate__())
+
+        subcls.__name__ = subcls.__qualname__ = fmt.format(base_class.__name__)
+        if attr_name is not None:
+            setattr(subcls, attr_name, base_class)
+        return subcls
+
+    return class_factory
+
+
+def _picklable_class_constructor(
+        base_cls, fmt, attr_name, _default_base_class, base_class):
```

Comment:
Surely there's a better name to use than `base_cls` and `base_class`?

Reply:
oops, the first one should have been mixin_class anyways...